### ISSUE TITLE
License exceptions are listed as "spdx_license_key"

### ISSUE THREAD

**Issue by sschuberth** [nexB/scancode-toolkit#1159]
I'm using (copied from JSON output)

```
  "scancode_version": "2.9.2.post308.ec97cb651",
  "scancode_options": {
    "input": "/home/seschube/Downloads/config.guess",
    "--copyright": true,
    "--json-pp": "a.json",
    "--license": true
  },
```

to scan https://github.com/eclipse/titan.core/blob/c07d711ed00f02ddfc66667952a64eb6da3a3fc6/mctr2/editline/libedit/config.guess, which yields

        {
          "key": "autoconf-exception-2.0",
          "score": 100.0,
          "short_name": "Autoconf exception to GPL 2.0 or later",
          "category": "Copyleft Limited",
          "owner": "Free Software Foundation (FSF)",
          "homepage_url": "",
          "text_url": "http://ac-archive.sourceforge.net/doc/copyright.html",
          "reference_url": "https://enterprise.dejacode.com/urn/urn:dje:license:autoconf-exception-2.0",
          "spdx_license_key": "Autoconf-exception-2.0",
          "spdx_url": "https://spdx.org/licenses/Autoconf-exception-2.0",
          "start_line": 9,
          "end_line": 27,
          "matched_rule": {
            "identifier": "gpl-2.0-plus_with_autoconf_2.RULE",
            "license_expression": "gpl-2.0-plus WITH autoconf-exception-2.0",
            "licenses": [
              "gpl-2.0-plus",
              "autoconf-exception-2.0"
            ]
          }
        },

However, "Autoconf-exception-2.0" is not a valid "spdx_license_key", it's an [exception](https://github.com/spdx/license-list-data/blob/v3.2/json/exceptions/Autoconf-exception-2.0.json). The correct license in this case would be [GPL-2.0-with-autoconf-exception](https://github.com/spdx/license-list-data/blob/v3.2/json/details/GPL-2.0-with-autoconf-exception.json).

**Comment by pombredanne** [comment]
@sschuberth there is only one data structure in ScanCode for licenses (and exceptions)... and there is an `is_exception` flag to tag licenses that are "exceptions"... this is different from the way SPDX uses two different data structures, one for licenses and one for exceptions.
So here IMHO the detection is correct (since `GPL-2.0-with-autoconf-exception` is now a deprecated license id at SPDX) with the reported expression: `"gpl-2.0-plus WITH autoconf-exception-2.0",`

What is missing is IMHO reporting the `is_exception` flag with each discrete license record, e.g. something like:

```
 {
      "key": "autoconf-exception-2.0",
.....
      "is_exception": true,
      "spdx_license_key": "Autoconf-exception-2.0",
      "spdx_url": "https://spdx.org/licenses/Autoconf-exception-2.0",
      "start_line": 9,
      "end_line": 27,
      "matched_rule": {
        "identifier": "gpl-2.0-plus_with_autoconf_2.RULE",
        "license_expression": "gpl-2.0-plus WITH autoconf-exception-2.0",
        "licenses": [
          "gpl-2.0-plus",
          "autoconf-exception-2.0"
        ]
      }
```

Note also that there is a bug at SPDX since https://spdx.org/licenses/Autoconf-exception-2.0 is a dead link... and this is mildly inconsistent since the URL prefix there is the same for licenses and exceptions but they use two different lists.

Now since we are on this topic, I am not too happy with the way licenses are reported now that we support expressions: we have a list of `license_expressions` and still a list of `licenses` that embed a repetition of the data for each license key and also repeats the detection information at the rule level.
 I feel this is a tad messy and verbose... would you agree?

**Comment by sschuberth** [comment]
> I feel this is a tad messy and verbose... would you agree?

I do agree it's verbose, but it's not too much data that gets duplicated, any maybe it's useful for tools that work with ScanCode's output but that do not support license expressions. I.e. parsing the list of licenses from the expression is a bit more involving than simply getting it from JSON. So I would not drop the `licenses` field. On the other hand, from this plain list the relation is unclear (`AND`, `OR`, `WITH`, etc.), so we also need the `license_expression`.

An alternative could be to turn `licenses` into a list of pairs, each pair consisting of the license id and the relation to the following license, if any. Or leave `licenses` as-is, and introduce a `license_relations` list with N-1 entries that contain the relations in order, and you'd get the license expression basically by zipping / interleaving the `licenses` and `license_relations` lists. But that's probably a bit unobvious.

**Comment by sschuberth** [comment]
> Note also that there is a bug at SPDX since https://spdx.org/licenses/Autoconf-exception-2.0 is a dead link...

Well, you omitted the `.html` extension. Or did you want to get the plain text of the exception?

**Comment by pombredanne** [comment]
@sschuberth thanks for the feedback, yes I wanted the plain text and that's an SPDX bug that I need to report

**Comment by sschuberth** [comment]
But even for licenses you need to append `.txt`, i.e. use https://spdx.org/licenses/0BSD.txt instead of https://spdx.org/licenses/0BSD.

**Comment by pombredanne** [comment]
Good point: yet the plain name without extension should link alright... and always was... so that's a bug in anycase

**Comment by sschuberth** [comment]
What still causes a bit of grief on my side is that the field is still called "spdx_*license*_key", but it still is no plain *license* in this case. Maybe it should be more generally called "spdx_key", but that would be a breaking change. It's just quibble after all and probably not worth the effort.

But another thing is that I feel exceptions should not live one the same level as licenses. You cannot have an exception without a license, hence a license should contain / refer to an exception, if any, so there should be a hierarchy between the two. Just my 2 cents.

**Comment by pombredanne** [comment]
> You cannot have an exception without a license, hence a license should contain / refer to an exception, if any, so there should be a hierarchy between the two. 

From a detection point of view, it happens that you can have an exception text found in a codebase as-is and that does not have a license with it (in the same file). That's weird, but it happens.

Also I do not think I could build a set of finite relationships between exceptions and licenses to build a reliable reference table...

But mostly, what I am not sure I grok is the benefit from handling the exception models differently or tracking extra references and/or relationships?

Exceptions are a "kind of" license in terms of attributes, so it makes sense that they share the same base definition as a data structure. The only difference is that they have a different type: "exceptions" or "license" and there are two ways to model this:
1. use a flag (what we have now) or a type field in a single data structure
2. use two data structures to capture the type in two data structures

I went with 1. which was much simpler as a code evolution and I do not see what would be the benefits of 2.

Also, the only place where there would be a need to know if a license is an "exception" is when we parse license expressions and only if we wanted to be strict about the fact that the right side of a `WITH` must be an exception. There is one place where expressions are parsed today (when detecting SPDX license identifiers) and there is another one coming up (for package manifests that use expressions)... but even there the parsing is not done in strict mode so there is no issues with a license being used in place of an exception in a `WITH` statement.  The only refinements I could see could be to report something extra in the license expression detection if this license expression is not valid (e.g. uses invalid types in a `WITH` expression.

**Comment by pombredanne** [comment]
@sschuberth any further comment on this?

**Comment by sschuberth** [comment]
> But mostly, what I am not sure I grok is the benefit from handling the exception models differently or tracking extra references and/or relationships?

I stumpled upon this because when generating NOTICE files with ORT, we look up the original license text via the SPDX license list, based on the `id` ScanCode reports. Previously, there was no easy way to distinguish license `id`s from exception `id`s, so I had to implement some [ugly work-around](https://github.com/heremaps/oss-review-toolkit/blob/6f0d4ae29a2c55ba217ec442d5ffa5786511aaf3/utils/src/main/kotlin/SpdxUtils.kt#L76).

As SPDX also treats exceptions differently in the sense that their data is located in a separate `exceptions` directory, I thought it would make sense to treat then as different things in ScanCode, too.

> I went with 1. which was much simpler as a code evolution and I do not see what would be the benefits of 2.

That's good enough for me, thanks. In the end what counts for me is that we can clearly distunguish license ids / findings / texts from exception ids / findings / texts.

> if we wanted to be strict about the fact that the right side of a `WITH` must be an exception.

I believe that would make sense, yes.
